### ISSUE TITLE
Need a new feature for administrators and moderators

### ISSUE THREAD

**Issue by StefOfficiel** [mastodon/mastodon#22424]
### Pitch

Returning a URL to a Mastodon page before redirecting to the actual site.

### Motivation

There is a real need to have all the links shared in a Toot which would initially redirect to "https://MyInstance.mastodon?link=https://external.link"

In the administration, a filter system could be created, which would warn the user of a fraudulent link or one that could harm them.

So, two scenarios:
The first, the url is detected as dangerous, then there is a warning message (customizable in Markdown for example) but which leaves the choice to the user to visit or not the site in question. Example "This link redirects you to example.tld which is a phishing site, return to a safe place or click here to visit example.tld anyway"
The second, the url is safe, so the redirection is done invisibly.

**Comment by rbairwell** [comment]
Reasons against this:
* Privacy (one of the main "Mastodon selling points"). I might not want my admin to know what I've clicked on.
* Security. You can't have a redirect like that as it'll be open for abuse (see "Open Redirection Attacks"), so the URL would need to either be "hidden" (the way Twitter does it on t.co and therefore requires database storage) or contain some sort of of verifiable hash linked to the instance.
* Overhead. In the Mastodon scale of things, having the server do "one more" http receive, database lookup and then transmitting of the redirect isn't much, but it's still something.
* Federation. Posts are sent as a filtered HTML block across the Fediverse. This change would either involve servers having to rewrite every single URL that passes through them OR have all links link back to the originating instance (which, again, goes against the Mastodon security/privacy "selling points")

For dangerous links, I would suggest just submitting them to [Google Safebrowsing](https://safebrowsing.google.com/safebrowsing/report_badware/?hl=en) , [Microsoft Security Intelligence](https://www.microsoft.com/en-us/wdsi/support/report-unsafe-site-guest), [PhishTank](https://www.phishtank.com/) - most of which can be done via sites such as [Phish.Report](https://phish.report/) . That means that non-Mastodon users get the security benefits as well (most browsers use either Google or Microsoft's filtering list) and there stands a better chance of the site actually being taken down.  Plus it has the "advantage" of not being "misused" by an instance admin (say, for a totally random example, a competitor website is growing and so you want to put fake "This link is malware" warnings on links to them...*cough* Twitter *cough*)

**Comment by StefOfficiel** [comment]
Hi,
I understand the arguments, and I think that we are wrong in the analysis, I will explain myself:

```
- Privacy (one of the main "Mastodon selling points"). I might not want my admin to know what I've clicked on.
```

Whatever happens, in the server logs we can already know if a link has been clicked...
The goal is not to track the user but to be able to inform him about a danger at the discretion of the administrator of the instance.

```
- Security. You can't have a redirect like that as it'll be open for abuse (see "Open Redirection Attacks"), so the URL would need to either be "hidden" (the way Twitter does it on t.co and therefore requires database storage) or contain some sort of of verifiable hash linked to the instance.
```

Yes, indeed, I see the potential security issue. Nothing prevents allowing only requests that come from the instance (Referrer-Policy Cross-Origin Strict).

```
je suggérerais de les soumettre à Google Safebrowsing, Microsoft Security Intelligence, PhishTank
```

On the other points, offering to use Google and Microsoft services when you want to respect the privacy of its users, cough... it's not really librist cough... cough...

Exactly, let's talk about Twitter! Twitter is openly an extreme right-wing network, do we want to promote this type of network? Should we condone that Elon Musk is blocking "John Mastodon" and that we should continue to feed the extreme right-wing?

Again, it's a tool for the administration of the instance, if the administrator decides that The Smurfs is dangerous for his users, then he will!
Because the other unconditional rule of Mastodon is that you join an instance by being in agreement with the rules of this one and if you are not or no longer in agreement with it, then you can migrate to an instance that suits us best or create your own instance!

Be that as it may, we need to give administrators more means to take their role seriously with the tool without encouraging them to tweak the code!

**Comment by ThisIsMissEm** [comment]
I thought we had a feature request for "Allow Moderators to apply Roadblocks to Links", however, I can't find it at present.

That's what this feature would be, such that only links that have been moderated in some way would be rewritten to go through the roadblock.

**Comment by ThisIsMissEm** [comment]
I'd perhaps suggest rewording this issue to be "Allow Moderators to apply roadblocks to Links", we'd also need an ability to report a link, which would require Polymorphic Reports (#26897)

**Comment by ClearlyClaire** [comment]
> I'd perhaps suggest rewording this issue to be "Allow Moderators to apply roadblocks to Links", we'd also need an ability to report a link, which would require Polymorphic Reports

(I don't think you've linked the correct PR/issue in this)
